### PR TITLE
const API

### PR DESCRIPTION
**Pull request by mciantyre**
The PR refactors the boot-gen crate. The 0.1 release required a build script to define the boot data structures, and to emit the memory as a byte array. In this new release, we provide a `const` API to generate the data structures in your embedded code. The goal was to make the library easier to use (no more build scripts), and easier for us to maintain and extend (no need to maintain a data structure writer, define a new writer for different data structures). The PR also updates documentation, and library organization.

See the CHANGELOG for migration tips, and the tests for an example of the new API.

### CODE REVIEW

**Comment by mciantyre** [comment]
See mciantyre/teensy4-rs#89 for a concise example of the new API.

**Comment by bfrog** [comment]
I like this change quite a lot, its nice to no longer need a build.rs for defining the flash config block! I'm impressed const fn's are able to provide this

File: LICENSE-APACHE [hunk 1]
```diff
@@ -186,7 +186,7 @@ APPENDIX: How to apply the Apache License to your work.
    same "printed page" as the copyright notice for easier
    identification within third-party archives.
 
-Copyright 2019 Ian McIntyre
+Copyright [yyyy] [name of copyright owner]
```
Comment (bfrog) [inline LICENSE-APACHE:189]: This seems like a possible mistake

File: LICENSE-MIT [hunk 2]
```diff
@@ -1,23 +1,21 @@
...
-publish, distribute, sublicense, and/or sell copies of
-the Software, and to permit persons to whom the Software
-is furnished to do so, subject to the following
-conditions:
+MIT License
 
-The above copyright notice and this permission notice
-shall be included in all copies or substantial portions
-of the Software.
+Copyright (c) 2019-2020 Ian McIntyre
```
Comment (bfrog) [inline LICENSE-MIT:3]: :+1:

File: src/flexspi.rs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+///         .cs_setup_time(0x02)
+///         .column_address_width(ColumnAddressWidth::OtherDevices)
+///         .device_mode_configuration(DeviceModeConfiguration::Disabled)
+///         .wait_time_cfg_commands(WaitTimeConfigurationCommands::new(40_000))
+///         .flash_size(SerialFlashRegion::A1, 0x0020_0000)
+///         .serial_clk_freq(SerialClockFrequency::MHz60)
+///         .serial_flash_pad_type(FlashPadType::Quad);
+///
+#[derive(Debug, Clone, Copy)]
+#[repr(C, packed)]
```
Comment (bfrog) [inline src/flexspi.rs:115]: It's nice that you are able to to use repr(C, packed) here to effectively name the various memory sections in the configuration block here

**Review by bfrog — APPROVED** [review]
Looks great, love the usage of the packed C struct to map to memory. Since it works I assume the byte ordering matches and things are gravy.

Nice!

File: LICENSE-APACHE (hunk 1 above)
Comment (mciantyre) [reply LICENSE-APACHE:189]: This one looks OK. I incorrectly filled in the license boilerplate within the license, which doesn't follow the license's recommendation. We're expected to replace the *after* we copy the notice to the top of the source file.

Comment (bfrog) [reply LICENSE-APACHE:189]: Ahhh

**Pull request merged**